### PR TITLE
infra(ci): Device Farm test splitting and runtime filter (iOS + Android)

### PR DESCRIPTION
**Pull request by tobi-legan**
## 🎯 What problem does this PR solve?

- Device Farm runs were unstable — all 17 LLM tests ran sequentially on a single device, causing resource exhaustion, crashes, and timeouts
- No mechanism to truly split test execution inside the app — WDIO `mochaOpts.grep` only filtered monitoring, not actual execution
- OIDC tokens expired during long Device Farm monitoring phases
- No crash detection — app crashes were only discovered at the end via timeout

### Workflow Changes (`integration-mobile-test-qvac-lib-infer-llamacpp-llm.yml`)
- **iOS**: 7-way test splitting (5 heavy + 2 light groups), each scheduled as a separate Device Farm run
- **Android**: 2-way test splitting (Group A: 9 tests, Group B: 8 tests), scheduled as parallel Device Farm runs
- Pool-based scheduling via `device-pool-arn` instead of per-device filters
- `aws_retry` helper for transient AWS API failures during monitoring
- OIDC credential refresh before monitor and download phases
- Crash detection via `queryAppState` in WDIO `before`/`afterTest` hooks — exits early if app state < 3

### Real Test Splitting (App-Level Filtering)
- Added `global.__shouldRunTest()` filter in `integration-runtime.cjs` — reads `testFilter.txt` from device Documents directory at runtime
- WDIO `before` hook pushes filter pattern to device via Appium `pushFile` before clicking "Run Automated Tests"
  - iOS: `@bundleId:documents/testFilter.txt`
  - Android: `@packageId/files/testFilter.txt`
- Each `run*Test()` wrapper checks filter and returns instantly for non-matching tests (zero resource cost)
- Updated `generate-mobile-integration-tests.js` to inject filter guards into generated wrappers

## 🧪 How was it tested?

- **iOS Device Farm**: 7 parallel runs — 6/7 consistently passing
  - 1 intermittent failure (LightB on iPhone 16 Pro) traced to pre-existing Bare Runtime bug where context overflow error handling causes `RangeError: Maximum call stack size exceeded` — not related to splitting changes
- **Android Device Farm**: 2-way split pending CI validation

Made with [Cursor](https://cursor.com)

### CODE REVIEW

**Review by olyasir — DISMISSED** [review]

**Review by gianni-cor — CHANGES_REQUESTED** [review]
## Requested Change: move test group definitions to `test-groups.json`

The iOS/Android group patterns are hardcoded in the workflow YAML. When a new test file is added and `integration.auto.cjs` is regenerated, the new test **silently gets skipped** in every Device Farm run because it won't match any group's filter regex — with no error or warning.

This also surfaces an existing bug: `runAfriquegemmaEdgeCasesTest` is in iOS Light A but doesn't exist in `integration.auto.cjs`.

**Fix:** define groups in a JSON file next to the test code, validate coverage in the generator, read patterns via `jq` in the workflow. Three changes:

---

### 1. Create `packages/qvac-lib-infer-llamacpp-llm/test/mobile/test-groups.json`

```json
{
  "ios": {
    "heavy1": ["runFinetuningPauseResumeTest"],
    "heavy2": ["runToolCallingTest"],
    "heavy3": ["runImageTest"],
    "heavy4": ["runOcrLightonTest"],
    "heavy5": ["runReasoningTest"],
    "lightA": [
      "runAfriquegemmaTranslationTest",
      "runApiBehaviorTest",
      "runBitnetTest",
      "runCacheStateMachineTest",
      "runConfigParametersTest",
      "runDynamicToolsTest"
    ],
    "lightB": [
      "runGenerationParamsTest",
      "runModelLoadingTest",
      "runMoeTest",
      "runMultiInstanceTest",
      "runSlidingContextTest",
      "runUtf8OutputTest"
    ]
  },
  "android": {
    "groupA": [
      "runAfriquegemmaTranslationTest",
      "runApiBehaviorTest",
      "runBitnetTest",
      "runCacheStateMachineTest",
      "runConfigParametersTest",
      "runDynamicToolsTest",
      "runFinetuningPauseResumeTest",
      "runGenerationParamsTest",
      "runImageTest"
    ],
    "groupB": [
      "runModelLoadingTest",
      "runMoeTest",
      "runMultiInstanceTest",
      "runOcrLightonTest",
      "runReasoningTest",
      "runSlidingContextTest",
      "runToolCallingTest",
      "runUtf8OutputTest"
    ]
  }
}
```

(Note: `runAfriquegemmaEdgeCasesTest` removed from iOS Light A — it doesn't exist in `integration.auto.cjs`.)

---

### 2. Add validation to `scripts/generate-mobile-integration-tests.js`

**a)** Add after the existing `const outputFile = …` line:

```js
const groupsFile = path.join(mobileDir, 'test-groups.json')
```

**b)** Add this function before `main()`:

```js
function validateGroups (functionNames) {
  if (!fs.existsSync(groupsFile)) {
    console.warn('[warn] test-groups.json not found — skipping split validation')
    return
  }
  const groups = JSON.parse(fs.readFileSync(groupsFile, 'utf-8'))
  const nameSet = new Set(functionNames)
  for (const [platform, splits] of Object.entries(groups)) {
    const covered = new Set(Object.values(splits).flat())
    const missing = functionNames.filter(n => !covered.has(n))
    const extra = [...covered].filter(n => !nameSet.has(n))
    if (missing.length) {
      throw new Error(
        '[' + platform + '] Tests not assigned to any group in test-groups.json:\n  ' +
        missing.join('\n  ') + '\nAdd them to a group in test/mobile/test-groups.json.'
      )
    }
    if (extra.length) {
      throw new Error(
        '[' + platform + '] test-groups.json references non-existent tests:\n  ' +
        extra.join('\n  ') + '\nRemove them or check for typos.'
      )
    }
  }
  console.log('Group coverage validated — all tests assigned for every platform.')
}
```

**c)** Update `main()` to call it:

```js
function main () {
  const files = getIntegrationFiles()
  if (files.length === 0) {
    throw new Error('No integration test files found inside ' + integrationDir)
  }
  const functionNames = files.map(toFunctionName)
  const content = buildFileContents(files)
  fs.writeFileSync(outputFile, content, 'utf8')
  console.log('Generated ' + outputFile + ' with ' + files.length + ' integration runners.')
  validateGroups(functionNames)
}
```

---

### 3. In the workflow, read groups from JSON instead of hardcoding

Add this line before the `if [ "${{ matrix.platform }}" == "Android" ]` block (right after the `make_split` function):

```bash
GROUPS_JSON="packages/qvac-lib-infer-llamacpp-llm/test/mobile/test-groups.json"
```

Replace the **Android** hardcoded assignments:

```bash
# replace these two lines:
#   ANDROID_GROUP_A="runAfriquegemmaTranslationTest|runApiBehaviorTest|…"
#   ANDROID_GROUP_B="runModelLoadingTest|runMoeTest|…"
# with:
ANDROID_GROUP_A=$(jq -r '.android.groupA | join("|")' "$GROUPS_JSON")
ANDROID_GROUP_B=$(jq -r '.android.groupB | join("|")' "$GROUPS_JSON")
```

Replace the **iOS** hardcoded assignments:

```bash
# replace all IOS_GREP_* assignments with:
IOS_GREP_HEAVY_1=$(jq -r '.ios.heavy1 | join("|")' "$GROUPS_JSON")
IOS_GREP_HEAVY_2=$(jq -r '.ios.heavy2 | join("|")' "$GROUPS_JSON")
IOS_GREP_HEAVY_3=$(jq -r '.ios.heavy3 | join("|")' "$GROUPS_JSON")
IOS_GREP_HEAVY_4=$(jq -r '.ios.heavy4 | join("|")' "$GROUPS_JSON")
IOS_GREP_HEAVY_5=$(jq -r '.ios.heavy5 | join("|")' "$GROUPS_JSON")
IOS_GREP_LIGHT_A=$(jq -r '.ios.lightA | join("|")' "$GROUPS_JSON")
IOS_GREP_LIGHT_B=$(jq -r '.ios.lightB | join("|")' "$GROUPS_JSON")
```

---

### Result

- Add a new test file → run `npm run test:mobile:generate` → generator **fails loudly** until you assign the test to a group in `test-groups.json`. No silent skipping.
- The workflow YAML never needs touching unless you change the **number** of groups.
- The ghost `runAfriquegemmaEdgeCasesTest` in iOS Light A is also fixed — the validator would reject it since that function doesn't exist.

File: .github/workflows/integration-mobile-test-qvac-lib-infer-llamacpp-llm.yml [hunk 1]
```diff
@@ -947,160 +948,189 @@ jobs:
...
+            echo "test_spec_arn=$SPEC_ARN_A" >> $GITHUB_OUTPUT
+            SPEC_ARN_B=$(upload_spec testspec-android-b.yml "Android-B")
+            echo "test_spec_arn_2=$SPEC_ARN_B" >> $GITHUB_OUTPUT
+          else
+            IOS_GREP_HEAVY_1="runFinetuningPauseResumeTest"
+            IOS_GREP_HEAVY_2="runToolCallingTest"
+            IOS_GREP_HEAVY_3="runImageTest"
+            IOS_GREP_HEAVY_4="runOcrLightonTest"
+            IOS_GREP_HEAVY_5="runReasoningTest"
+            IOS_GREP_LIGHT_A="runAfriquegemmaEdgeCasesTest|runAfriquegemmaTranslationTest|runApiBehaviorTest|runBitnetTest|runCacheStateMachineTest|runConfigParametersTest|runDynamicToolsTest"
```
Comment (gianni-cor) [inline .github/workflows/integration-mobile-test-qvac-lib-infer-llamacpp-llm.yml]: **Issue: ghost test name**

`runAfriquegemmaEdgeCasesTest` doesn't exist in `integration.auto.cjs` — it was never generated (there's no corresponding `afriquegemma-edge-cases.test.js` in `test/integration/`). It won't cause a runtime error (mocha grep just won't match anything), but it's dead weight in the filter pattern and inflates the apparent test count.

Remove it from Light A, or if the test file is planned, add it to `test/integration/` and regenerate.

File: .github/workflows/integration-mobile-test-qvac-lib-infer-llamacpp-llm.yml [hunk 2]
```diff
@@ -947,160 +948,189 @@ jobs:
...
+            local ARN=$(echo $RESP | jq -r '.upload.arn')
+            curl -T "$spec_file" "$URL"
+            local MAX=20; local A=1
+            while [ $A -le $MAX ]; do
+              local ST=$(aws devicefarm get-upload --arn "$ARN" --query "upload.status" --output text)
+              if [ "$ST" = "SUCCEEDED" ]; then echo "$ARN"; return 0; fi
+              if [ "$ST" = "FAILED" ]; then echo "$label spec upload failed" >&2; exit 1; fi
+              sleep 5; A=$((A + 1))
+            done
+            echo "$ARN"
```
Comment (gianni-cor) [inline .github/workflows/integration-mobile-test-qvac-lib-infer-llamacpp-llm.yml]: **Bug: `upload_spec()` doesn't fail on timeout**

If all 20 poll attempts exhaust without reaching `SUCCEEDED` or `FAILED`, the function falls through and echoes the ARN as if the upload succeeded. The caller then proceeds to schedule a Device Farm run with a potentially incomplete/broken spec — which will fail cryptically downstream.

Replace this fallthrough with an explicit failure:

```bash
            echo "$label spec upload timed out after $MAX attempts" >&2; exit 1
```

File: packages/qvac-lib-infer-llamacpp-llm/scripts/generate-mobile-integration-tests.js [hunk 3]
```diff
@@ -32,18 +32,24 @@ function buildFileContents (files) {
...
   lines.push('// AUTO-GENERATED FILE. Run `npm run test:mobile:generate` to update.')
   lines.push('// Each function mirrors a single file under test/integration/.')
+  lines.push('// Functions are invoked dynamically by the mobile test runner framework.')
   lines.push('')
   lines.push('/* global runIntegrationModule */')
   lines.push('')
 
+  lines.push('/* global __shouldRunTest */')
+  lines.push('')
+  lines.push('const __FILTERED = { modulePath: \'filtered\', summary: { total: 0, passed: 0, failed: 0 } }')
```
Comment (gianni-cor) [inline packages/qvac-lib-infer-llamacpp-llm/scripts/generate-mobile-integration-tests.js:43]: **Nit: return shape mismatch**

`__FILTERED` returns `{ modulePath, summary }` (an object), but `runIntegrationModule()` returns a plain string (`modulePath` or `'missing'`). If the mobile test runner consumes these return values for result reporting, it would need to handle both shapes.

Worth confirming the consumer can deal with both types. If it can, this is fine — just flagging the asymmetry.

File: .github/workflows/integration-mobile-test-qvac-lib-infer-llamacpp-llm.yml [hunk 4]
```diff
@@ -947,160 +948,189 @@ jobs:
...
+          }
+
+          # For each split: inject mocha grep AND replace __TEST_FILTER__ so
+          # the before-hook pushes a testFilter.txt the app reads at runtime.
+          # This ensures the app only executes matching tests (real splitting).
+          make_split() {
+            local pattern="$1" output="$2"
+            local cfg
+            cfg=$(echo "$WDIO_CONFIG" | sed "s/timeout:1800000}/timeout:1800000,grep:\"$pattern\"}/")
```
Comment (gianni-cor) [inline .github/workflows/integration-mobile-test-qvac-lib-infer-llamacpp-llm.yml]: **Fragile: `sed` injection depends on exact config structure**

This relies on the literal substring `timeout:1800000}` existing verbatim in the minified WDIO config. If the timeout value changes, or the config keys get reordered, the `sed` won't match and the `grep` option silently won't be injected — the spec would run **all** tests instead of just the intended split group, defeating the whole purpose.

The `__TEST_FILTER__` sentinel on the next line is much more robust. Consider the same approach for mocha grep — add a `__MOCHA_GREP__` placeholder in the WDIO_CONFIG template string and replace it here:

```suggestion
            cfg=$(echo "$WDIO_CONFIG" | sed "s#__MOCHA_GREP__#$pattern#")
```

Then in both WDIO_CONFIG definitions above, replace `timeout:1800000}` with `timeout:1800000,grep:"__MOCHA_GREP__"}` (or `grep:""` as a no-op default if you prefer).

**Review by gianni-cor — APPROVED** [review]
All review items addressed cleanly in `74688e6`:

- **test-groups.json** created with correct 17-test coverage across both platforms
- **validateGroups()** added to generator — fails loudly on missing/extra test names
- **`__MOCHA_GREP__` sentinel** replaces fragile `sed` on `timeout:1800000}`
- **`upload_spec()` timeout** now exits explicitly instead of falling through
- **`runIntegrationModule` return shape** aligned with `__FILTERED` (consistent object shape)
- **Ghost `runAfriquegemmaEdgeCasesTest`** removed from iOS Light A
- Workflow reads groups via `jq` from JSON — nice use of `${{ env.ADDON_WORKDIR }}` for the path

LGTM.

**Review by olyasir — DISMISSED** [review]

**Review by gianni-cor — DISMISSED** [review]

**Comment by Victor-Rodzko** [comment]
Approved

@tobi-legan In the future, it would be cool to move the test-groups.json file into the test definition. That is, to somehow mark it within the test:
```
skip: ['iOS', 'Android', 'Linux'],
group: 'heavy1'
...
```

**Review by Victor-Rodzko — APPROVED** [review]

**Review by olyasir — APPROVED** [review]
